src/move/research/prover-examples/sources/value_holder.move

```diff
@@ -0,0 +1,180 @@
+    ):
+        exists<ValueHolder>(account_address)
+            && global<ValueHolder>(account_address).value
+                == global<ValueHolderManifest>(@prover_examples).value;
+
+    /// Ensure that on the creation of a value holder, the account address is originally absent
+    /// from the manifest then gets added, and that the value in the value holder is the same as
+    /// the value in the manifest.
+    invariant update[suspendable] forall account_address: address where !old(
+        exists<ValueHolder>(account_address)
+    ) && exists<ValueHolder>(account_address):
+        contains(
+            global<ValueHolderManifest>(@prover_examples).account_addresses,
+            account_address
+        )
+            && old(
+                !contains(
+                    global<ValueHolderManifest>(@prover_examples).account_addresses,
+                    account_address
+                )
+            )
+            && global<ValueHolder>(account_address).value
+                == global<ValueHolderManifest>(@prover_examples).value;
+
+    /// Ensure that if an account address is originally absent from the manifest but is then
+    /// added, a value holder is created at the corresponding address with the correct value.
+    invariant update[suspendable] forall account_address: address where !old(
+        contains(
+            global<ValueHolderManifest>(@prover_examples).account_addresses,
+            account_address
+        )
+    ) && contains(
+        global<ValueHolderManifest>(@prover_examples).account_addresses,
+        account_address
+    ):
+        exists<ValueHolder>(account_address)
+            && old(!exists<ValueHolder>(account_address))
+            && global<ValueHolder>(account_address).value
+                == global<ValueHolderManifest>(@prover_examples).value;
+
+    /*
+        /// Ensure that if the value in the manifest is updated, all value holders are updated.
+        invariant update[suspendable] forall new_value: u8, account_address: address where old(
+            exists<ValueHolderManifest>(@prover_examples)
+        ) && old(exists<ValueHolder>(account_address)):
+            global<ValueHolder>(account_address).value == new_value;
+    */
+
+    public fun init_value_holder(account: &signer) acquires ValueHolderManifest {
+        let value_holder_manifest_ref_mut = &mut ValueHolderManifest[@prover_examples];
+        move_to(
+            account,
+            ValueHolder { value: value_holder_manifest_ref_mut.value }
+        );
+        value_holder_manifest_ref_mut.account_addresses.push_back(
+            signer::address_of(account)
+        );
+    }
+
+    public fun update_value(new_value: u8) acquires ValueHolder, ValueHolderManifest {
+        let value_holder_manifest_ref_mut = &mut ValueHolderManifest[@prover_examples];
+        value_holder_manifest_ref_mut.value = new_value;
+        value_holder_manifest_ref_mut.account_addresses.for_each_ref(|account_address_ref| {
```

Comment:
The prover does not support inline functions for now so you need to expand `for_each_ref` as the while loop and add the corresponding loop invariant that states that `global<ValueHolder>(account_address).value == global<ValueHolderManifest>(@prover_examples).value` for each account address in value_holder_manifest.account_addresses.

Reply:
@rahxephon89 in f5bd875 I added loop invariants, but now they are erroring out:

```sh
error: global memory invariant does not hold
   ┌─ /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:54:5
   │  
54 │ ╭     invariant [suspendable] forall account_address: address where exists<ValueHolder>(
55 │ │         account_address
56 │ │     ):
57 │ │         exists<ValueHolderManifest>(@prover_examples)
   · │
62 │ │             && global<ValueHolder>(account_address).value
63 │ │                 == global<ValueHolderManifest>(@prover_examples).value;
   │ ╰───────────────────────────────────────────────────────────────────────^
   │  
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:129: update_value
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:41
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:129: update_value
   =         new_value = 34u8
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:131: update_value
   =         value_holder_manifest_ref_mut =
   =           &value_holder.ValueHolderManifest{
   =             account_addresses = vector{0xabc, 0xabf},
   =             value = 35u8}
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:132: update_value
   =         $t3 = &35u8
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:135: update_value
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:19
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:129: update_value
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:135: update_value
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:49
   =     at /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:54

error: verification out of resources/timeout (global timeout set to 40s)
    ┌─ /Users/yours-truly/repos/econia-v5/src/move/research/prover-examples/sources/value_holder.move:118:5
    │  
118 │ ╭     public fun init_value_holder(account: &signer) acquires ValueHolderManifest {
119 │ │         let value_holder_manifest_ref_mut = &mut ValueHolderManifest[@prover_examples];
120 │ │         move_to(
121 │ │             account,
    · │
126 │ │         );
127 │ │     }
    │ ╰─────^

{
  "Error": "Move Prover failed: exiting with verification errors"
}
```

It looks like there are two errors:

1. Timeout
2. This condition appears to be violated, but I am still not able to make out from the trace why not: https://github.com/econia-labs/econia-v5/blob/f5bd8758608b4055e3d9765d0e80f784d8a196a8/src/move/research/prover-examples/sources/value_holder.move#L52-L63

Is there something I'm missing?